<a href="https://colab.research.google.com/github/Maurice912/DATA_preprocessing_project/blob/main/notebooks/Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Step 1 – Load and Explore Kaggle Dataset




In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
df = pd.read_csv('/content/drive/MyDrive/Data_preprocessing/population_data.csv')

In [3]:
# Preview data
print("Preview of the data:")
display(df.head())

# Show dataset shape
print(f"\nDataset shape: {df.shape[0]} rows and {df.shape[1]} columns")

# Show dataset info
print("\nDataset info:")
df.info()

# Summary statistics
print("\nSummary statistics:")
display(df.describe())

# Column names
print("\nColumn names:")
print(df.columns.tolist())

Preview of the data:


id Country (or dependency)  Population 2025 Yearly Change  Net Change  \
0   1                   India       1463865525         0.89%    12929734   
1   2                   China       1416096094        −0.23%  −3,225,184   
2   3           United States        347275807         0.54%     1849236   
3   4               Indonesia        285721236         0.79%     2233305   
4   5                Pakistan        255219554         1.57%     3950390   

   Density (P/Km²)  Land Area (Km²) Migrants (net)  Fert. Rate  Median Age  \
0              492          2973190       −495,753        1.94        28.8   
1              151          9388211       −268,126        1.02        40.1   
2               38          9147420        1230663        1.62        38.5   
3              158          1811570        −39,509        2.10        30.4   
4              331           770880     −1,235,336        3.50        20.6   

  Urban Pop % World Share  
0       37.1%      17.78%  
1       67.5%      17.20%  
2       82.8%       4.22%  
3       59.6%       3.47%  
4       34.4%       3.10%


Dataset shape: 233 rows and 12 columns

Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 233 entries, 0 to 232
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       233 non-null    int64  
 1   Country (or dependency)  233 non-null    object 
 2   Population 2025          233 non-null    int64  
 3   Yearly Change            233 non-null    object 
 4   Net Change               233 non-null    object 
 5   Density (P/Km²)          233 non-null    int64  
 6   Land Area (Km²)          233 non-null    int64  
 7   Migrants (net)           233 non-null    object 
 8   Fert. Rate               233 non-null    float64
 9   Median Age               233 non-null    float64
 10  Urban Pop %              210 non-null    object 
 11  World Share              233 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 22.0+ KB

Summary statistics:


id  Population 2025  Density (P/Km²)  Land Area (Km²)  \
count  233.000000     2.330000e+02       233.000000     2.330000e+02   
mean   117.000000     3.532080e+07       487.673820     5.583427e+05   
std     67.405489     1.391905e+08      2412.937646     1.694270e+06   
min      1.000000     5.010000e+02         0.000000     0.000000e+00   
25%     59.000000     4.663300e+05        38.000000     2.830000e+03   
50%    117.000000     5.731206e+06        98.000000     8.220000e+04   
75%    175.000000     2.407458e+07       256.000000     4.103400e+05   
max    233.000000     1.463866e+09     25732.000000     1.637687e+07   

       Fert. Rate  Median Age  
count  233.000000  233.000000  
mean     2.306094   31.896996  
std      1.134316    9.860936  
min      0.690000   14.500000  
25%      1.470000   22.900000  
50%      1.940000   32.800000  
75%      2.950000   39.800000  
max      5.940000   57.400000


Column names:
['id', 'Country (or dependency)', 'Population 2025', 'Yearly Change', 'Net Change', 'Density (P/Km²)', 'Land Area (Km²)', 'Migrants (net)', 'Fert. Rate', 'Median Age', 'Urban Pop %', 'World Share']


In [4]:
# Check for missing values
print("\nMissing values per column:")
print(df.isna().sum())

# Check for duplicates
print(f"\nNumber of duplicate rows: {df.duplicated().sum()}")

# Quick look at data types and non-numeric columns
print("\nNon-numeric columns:")
print(df.select_dtypes(exclude='number').columns.tolist())



Missing values per column:
id                          0
Country (or dependency)     0
Population 2025             0
Yearly Change               0
Net Change                  0
Density (P/Km²)             0
Land Area (Km²)             0
Migrants (net)              0
Fert. Rate                  0
Median Age                  0
Urban Pop %                23
World Share                 0
dtype: int64

Number of duplicate rows: 0

Non-numeric columns:
['Country (or dependency)', 'Yearly Change', 'Net Change', 'Migrants (net)', 'Urban Pop %', 'World Share']


**Exploration Summary**

The dataset contains 233 rows and 12 columns, each representing a country or dependency with projected population and demographic indicators for 2025. All columns are complete except for Urban Pop %, which has 23 missing values. There are no duplicate entries. Key variables include Population 2025, Fertility Rate, Median Age, and Urban Pop %. Several columns contain non-numeric symbols such as “%” or commas, which will need to be cleaned before analysis. Overall, the dataset is well-structured and provides a clear overview of global population trends, forming a strong base for integration with life expectancy and education data in later steps.









# Step 2 Web Scraping

In [5]:
import requests

url = "https://www.worldometers.info/demographics/life-expectancy/"
headers = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/124.0.0.0 Safari/537.36"
    )
}

resp = requests.get(url, headers=headers, timeout=30)
print("Status:", resp.status_code, "| HTML length:", len(resp.text))

# keep the HTML for the next step
html = resp.text


Status: 200 | HTML length: 189368


In [6]:
import pandas as pd

tables = pd.read_html(html)  # use the HTML we fetched in 6.1
print("Number of tables found:", len(tables))
for i, t in enumerate(tables[:3]):  # quick peek at the first 3
    print(f"\n--- Table {i} preview ---")
    display(t.head(5))


Number of tables found: 1

--- Table 0 preview ---


/tmp/ipython-input-4089111155.py:3: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(html)  # use the HTML we fetched in 6.1


Rank           Country  Life Expectancy (both sexes)  \
0     1         Hong Kong                         85.77   
1     2             Japan                         85.00   
2     3       South Korea                         84.53   
3     4  French Polynesia                         84.31   
4     5       Switzerland                         84.23   

   Females Life Expectancy  Males Life Expectancy  
0                    88.39                  83.10  
1                    88.03                  81.99  
2                    87.40                  81.44  
3                    86.74                  82.03  
4                    86.06                  82.34

In [7]:
# Copy the full table so we keep every column
df_life = tables[0].copy()
df_life.columns = [str(c).strip() for c in df_life.columns]

print("Columns in scraped table:")
print(df_life.columns.tolist())

# Preview the first few rows
display(df_life.head())


Columns in scraped table:
['Rank', 'Country', 'Life Expectancy (both sexes)', 'Females Life Expectancy', 'Males Life Expectancy']


Rank           Country  Life Expectancy (both sexes)  \
0     1         Hong Kong                         85.77   
1     2             Japan                         85.00   
2     3       South Korea                         84.53   
3     4  French Polynesia                         84.31   
4     5       Switzerland                         84.23   

   Females Life Expectancy  Males Life Expectancy  
0                    88.39                  83.10  
1                    88.03                  81.99  
2                    87.40                  81.44  
3                    86.74                  82.03  
4                    86.06                  82.34

In [8]:
import re

# 1️⃣ Keep original scraped table safe
df_life_raw = df_life.copy()

# 2️⃣ Standardize column names
df_life.columns = (
    df_life.columns
    .str.strip()           # remove extra spaces
    .str.lower()           # lowercase
    .str.replace(' ', '_') # replace spaces with underscores
)

print("Standardized column names:")
print(df_life.columns.tolist())

# 3️⃣ Convert numeric columns (leave 'country' as text)
# define numeric columns using your actual column names
num_cols = ['rank', 'life_expectancy_(both_sexes)', 'females_life_expectancy', 'males_life_expectancy']

# only keep ones that actually exist (in case the site changes)
num_cols = [c for c in num_cols if c in df_life.columns]

# convert them
df_life[num_cols] = df_life[num_cols].apply(
    lambda s: pd.to_numeric(
        s.astype(str).str.replace(r'[^\d\.\-]', '', regex=True),
        errors='coerce'
    )
)

# 4️⃣ Quick preview
print(f"\nShape: {df_life.shape[0]} rows × {df_life.shape[1]} columns")
display(df_life.head())

# 5️⃣ Confirm data types
print("\nData types:")
print(df_life.dtypes)

# ✅ At this point:
# - Column names are clean and Python-friendly
# - Country names are preserved
# - Numeric columns are properly typed
# - Dataset is ready for checking missing values or duplicates


Standardized column names:
['rank', 'country', 'life_expectancy_(both_sexes)', 'females_life_expectancy', 'males_life_expectancy']

Shape: 201 rows × 5 columns


rank           country  life_expectancy_(both_sexes)  \
0     1         Hong Kong                         85.77   
1     2             Japan                         85.00   
2     3       South Korea                         84.53   
3     4  French Polynesia                         84.31   
4     5       Switzerland                         84.23   

   females_life_expectancy  males_life_expectancy  
0                    88.39                  83.10  
1                    88.03                  81.99  
2                    87.40                  81.44  
3                    86.74                  82.03  
4                    86.06                  82.34


Data types:
rank                              int64
country                          object
life_expectancy_(both_sexes)    float64
females_life_expectancy         float64
males_life_expectancy           float64
dtype: object


# Web Scraping Education Index Plus Cleaning

In [9]:
import requests
import pandas as pd
from io import StringIO

# --- 1) Fetch HTML (with headers to avoid 403) ---
url = "https://worldpopulationreview.com/country-rankings/education-index-by-country"
headers = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/124.0.0.0 Safari/537.36"
    )
}
resp = requests.get(url, headers=headers, timeout=30)
resp.raise_for_status()
html = resp.text
print("Status:", resp.status_code, "| HTML length:", len(html))

# --- 2) Parse tables (simple path) ---
tables = pd.read_html(StringIO(html))
print("Number of tables found:", len(tables))

# Use the first table (usually the main one)
df_edu = tables[0].copy()

# --- 3) Standardize column names & quick preview ---
df_edu.columns = [str(c).strip().lower().replace(' ', '_') for c in df_edu.columns]
print("Columns in scraped table:", df_edu.columns.tolist())
display(df_edu.head())

# --- 4) (Optional) Convert numeric-like columns to numeric ---
# Adjust names below if they differ on the page
num_cols = [c for c in df_edu.columns if 'index' in c or 'score' in c or c in ('rank',)]
if num_cols:
    df_edu[num_cols] = df_edu[num_cols].apply(
        lambda s: pd.to_numeric(s.astype(str).str.replace(r'[^\d\.\-]', '', regex=True), errors='coerce')
    )

# Keep a raw copy if you plan further cleaning
df_edu_raw = df_edu.copy()


Status: 200 | HTML length: 733978
Number of tables found: 1
Columns in scraped table: ['unnamed:_0', 'country', 'education_index_2022_(0-1.0)â\x86\x93']


unnamed:_0      country  education_index_2022_(0-1.0)↓
0         NaN    Australia                             1.01
1         NaN      Iceland                             0.99
2         NaN  New Zealand                             0.98
3         NaN      Germany                             0.96
4         NaN      Denmark                             0.96

In [12]:
print(df_edu.shape)
print(df_edu.columns.tolist())
df_edu.info()
display(df_edu.head())
display(df_edu.describe())


(192, 3)
['unnamed:_0', 'country', 'education_index_2022_(0-1.0)â\x86\x93']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 192 entries, 0 to 191
Data columns (total 3 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   unnamed:_0                       0 non-null      float64
 1   country                          192 non-null    object 
 2   education_index_2022_(0-1.0)↓  192 non-null    float64
dtypes: float64(2), object(1)
memory usage: 4.6+ KB


unnamed:_0      country  education_index_2022_(0-1.0)↓
0         NaN    Australia                             1.01
1         NaN      Iceland                             0.99
2         NaN  New Zealand                             0.98
3         NaN      Germany                             0.96
4         NaN      Denmark                             0.96

unnamed:_0  education_index_2022_(0-1.0)↓
count         0.0                       192.000000
mean          NaN                         0.679219
std           NaN                         0.179239
min           NaN                         0.240000
25%           NaN                         0.567500
50%           NaN                         0.700000
75%           NaN                         0.820000
max           NaN                         1.010000

# Data exploration of education index Summary
- This dataset includes 192 countries with their 2022 Education Index scores.
Values range from 0.24 to 1.01 (average = 0.68), showing wide variation in global education levels.
All country entries are complete, but one unused column will be removed and column names will be cleaned before merging with the other datasets.

In [13]:
# Drop any columns that start with 'unnamed'
df_edu = df_edu.loc[:, ~df_edu.columns.str.contains('^unnamed', case=False)]

print("Columns after dropping unnamed ones:")
print(df_edu.columns.tolist())


Columns after dropping unnamed ones:
['country', 'education_index_2022_(0-1.0)â\x86\x93']


In [14]:
# Fix messy encoding and standardize the education column name
df_edu.columns = (
    df_edu.columns
    .str.encode('ascii', 'ignore')     # remove non-ASCII characters like â\x86\x93
    .str.decode('utf-8')
    .str.strip()
    .str.lower()
    .str.replace(r'\s+', '_', regex=True)
    .str.replace('[^0-9a-zA-Z_]', '', regex=True)
)

print("Cleaned column names:")
print(df_edu.columns.tolist())


Cleaned column names:
['country', 'education_index_2022_010']
